dusty/systems/virtualbox/__init__.py

```diff
@@ -199,29 +200,55 @@ def get_docker_vm_ip():
             spec = line.split('=')[1].strip('"')
             name, protocol, host, host_port, target, target_port = spec.split(',')
             if name == 'ssh' and protocol == 'tcp' and target_port == '22':
-                ssh_port = host_port
-                break
-
-    if ssh_port:
-        # Now we can SSH to the VM on localhost:ssh_port. We do this to
-        # get inside the VM and determine the VM's local address on
-        # the virtualized network. This IP is what we're ultimately
-        # trying to return from this function.
-        ip_addr = check_output_demoted(['ssh', '-o', 'StrictHostKeyChecking=no',
-                                        '-o', 'UserKnownHostsFile=/dev/null',
-                                        '-i', _vm_key_path(), '-p', ssh_port,
-                                        'docker@127.0.0.1', 'ip addr show dev eth1'])
-        for line in ip_addr.splitlines():
-            line = line.strip()
-            if line.startswith('inet') and not line.startswith('inet6'):
-                ip = line.split(' ')[1].split('/')[0]
-                return ip
-
-    # If all our crazy hacks up there don't work, we can still get the IP
-    # from Machine directly. The reason we don't do this normally is
-    # because it's really slow, taking around 600ms on my fast machine.
-    logging.warning('Could not get IP the fast way, falling back to Docker Machine')
-    return check_output_demoted(['docker-machine', 'ip', constants.VM_MACHINE_NAME]).rstrip()
+                return host_port
+    raise ValueError('Could not determine localhost port for SSH forwarding')
+
+def _get_host_only_mac_address():
+    # Get the number of the host-only adapter
+    vm_config = _get_vm_config()
+    for line in vm_config:
+        if line.startswith('hostonlyadapter'):
+            adapter_number = int(line[15:16])
+            break
+    else:
+        raise ValueError('No host-only adapter is defined for the Dusty VM')
+
+    for line in vm_config:
+        if line.startswith('macaddress{}'.format(adapter_number)):
+            return line.split('=')[1].strip('"').lower()
+    raise ValueError('Could not find MAC address for adapter number {}'.format(adapter_number))
+
+def _ip_for_mac_from_ip_addr(ip_addr, target_mac):
+    return_next_ip = False
+    for line in ip_addr.splitlines():
+        line = line.strip()
+        if line.startswith('link/ether'):
+            line_mac = line.split(' ')[1].replace(':', '')
+            if line_mac == target_mac:
+                return_next_ip = True
+        elif return_next_ip and line.startswith('inet') and not line.startswith('inet6'):
+            ip = line.split(' ')[1].split('/')[0]
+            return ip
```

Comment:
there's no easy way to ask a machine to give you the IP assigned to the device with a given MAC address. this is parsing the output of `ip addr show`, which contains both of these things, to determine one from the other